File: src/plugins/xp.ts
import { IMatrixClient, Plugin, CommandDef, MessageContext } from "./base";
import { getDb } from "../db";
import { xpToLevel, xpForNextLevel, progressBar } from "../utils/parser";
import logger from "../utils/logger";

const XP_PER_MESSAGE = 10;
const XP_COOLDOWN_MS = 30_000;

const LEVEL_UP_MESSAGES = [
  "🔔✨ POWER UP! {user} hit Level {level}! Bell combo complete! ✨🔔",
  "🌟 {user} grabbed the golden bell — Level {level}! Speed up! 🐝💨",
  "⚡ Level {level}! {user} is fully powered up! Options attached! 🛸🛸",
  "🎮 {user} evolved to Level {level}! Twin attack ready! 👊👊",
  "🏆 STAGE CLEAR! {user} advanced to Level {level}! 🎆🎆🎆",
  "🔔🔔🔔 {user} collected enough bells for Level {level}! Shield activated! 🛡️",
  "📢 Level {level} unlocked! {user} picked up the megaphone power-up! 🔊💥",
  "⚡ {user} just warped to Level {level}! Laser mode engaged! 🔴🔴🔴",
  "🎰 Bonus round! {user} reached Level {level}! 3-way shot acquired! 🔱",
  "⚡ {user} powered through to Level {level}! Ripple laser online! 〰️💫",
];

export class XpPlugin extends Plugin {
  private cooldowns = new Map<string, number>();

  constructor(client: IMatrixClient) {
    super(client);
  }

  get name() {
    return "xp";
  }

  get commands(): CommandDef[] {
    return [
      { name: "rank", description: "Show your XP, level, and progress", usage: "!rank [@user]" },
      { name: "leaderboard", description: "Top users by XP", usage: "!leaderboard [n]" },
    ];
  }

  /**
   * Public method so other plugins (rep, wotd) can grant bonus XP.
   */
  grantXp(userId: string, roomId: string, amount: number, reason: string): void {
    const db = getDb();

    // Grab old level before granting
    const before = db.prepare(`SELECT level FROM users WHERE user_id = ? AND room_id = ?`).get(userId, roomId) as
      | { level: number }
      | undefined;
    const oldLevel = before?.level ?? 0;

    db.prepare(`
      INSERT INTO users (user_id, room_id, xp, level)
      VALUES (?, ?, ?, 1)
      ON CONFLICT(user_id, room_id) DO UPDATE SET
        xp = xp + ?,
        level = ?
    `).run(
      userId,
      roomId,
      amount,
      amount,
      // We recalculate level from the new XP total via a subquery-style approach.
      // Since better-sqlite3 is sync, we do a two-step:
      0 // placeholder — we fix level right after
    );

    // Fix level based on actual XP
    const row = db.prepare(`SELECT xp FROM users WHERE user_id = ? AND room_id = ?`).get(userId, roomId) as
      | { xp: number }
      | undefined;
    let newLevel = oldLevel || 1;
    if (row) {
      newLevel = xpToLevel(row.xp);
      db.prepare(`UPDATE users SET level = ? WHERE user_id = ? AND room_id = ?`).run(newLevel, userId, roomId);
    }

    db.prepare(`INSERT INTO xp_log (user_id, room_id, amount, reason) VALUES (?, ?, ?, ?)`).run(
      userId,
      roomId,
      amount,
      reason
    );

    logger.debug(`Granted ${amount} XP to ${userId} in ${roomId}: ${reason}`);

    // Announce level up
    if (newLevel > oldLevel && oldLevel > 0) {
      const template = LEVEL_UP_MESSAGES[Math.floor(Math.random() * LEVEL_UP_MESSAGES.length)];
      const msg = template.replace("{user}", userId).replace("{level}", String(newLevel));
      this.sendMessage(roomId, msg).catch((err) => {
        logger.error(`Failed to announce level up: ${err}`);
      });
    }
  }

  async onMessage(ctx: MessageContext): Promise<void> {
    // Passive XP grant with cooldown
    this.handlePassiveXp(ctx.sender, ctx.roomId);

    // Commands
    if (this.isCommand(ctx.body, "rank")) {
      await this.handleRank(ctx);
    } else if (this.isCommand(ctx.body, "leaderboard")) {
      await this.handleLeaderboard(ctx);
    }
  }

  private handlePassiveXp(userId: string, roomId: string): void {
    const key = `${userId}:${roomId}`;
    const now = Date.now();
    const lastGrant = this.cooldowns.get(key) ?? 0;

    if (now - lastGrant < XP_COOLDOWN_MS) return;

    this.cooldowns.set(key, now);
    this.grantXp(userId, roomId, XP_PER_MESSAGE, "message");
  }

  private async handleRank(ctx: MessageContext): Promise<void> {
    const args = this.getArgs(ctx.body, "rank");
    const targetUser = args.startsWith("@") ? args.split(/\s/)[0] : ctx.sender;

    const db = getDb();
    const row = db
      .prepare(`SELECT xp, level, rep FROM users WHERE user_id = ? AND room_id = ?`)
      .get(targetUser, ctx.roomId) as { xp: number; level: number; rep: number } | undefined;

    if (!row) {
      await this.sendReply(ctx.roomId, ctx.eventId, `No data found for ${targetUser}.`);
      return;
    }

    const progress = xpForNextLevel(row.xp);
    const bar = progressBar(progress.current, progress.needed);

    await this.sendMessage(
      ctx.roomId,
      `${targetUser}\nLevel ${row.level} | ${row.xp} XP | ${row.rep} rep\n${bar} ${progress.current}/${progress.needed} to next level`
    );
  }

  private async handleLeaderboard(ctx: MessageContext): Promise<void> {
    const args = this.getArgs(ctx.body, "leaderboard");
    const limit = Math.min(Math.max(parseInt(args) || 10, 1), 25);

    const db = getDb();
    const rows = db
      .prepare(`SELECT user_id, xp, level FROM users WHERE room_id = ? ORDER BY xp DESC LIMIT ?`)
      .all(ctx.roomId, limit) as { user_id: string; xp: number; level: number }[];

    if (rows.length === 0) {
      await this.sendReply(ctx.roomId, ctx.eventId, "No leaderboard data yet.");
      return;
    }

    const lines = rows.map((r, i) => `${i + 1}. ${r.user_id} — Lv${r.level} (${r.xp} XP)`);
    await this.sendMessage(ctx.roomId, `Leaderboard (Top ${rows.length}):\n${lines.join("\n")}`);
  }
}

File: internal/plugin/xp.go
package plugin

import (
	"database/sql"
	"fmt"
	"log/slog"
	"math/rand"
	"strings"
	"sync"
	"time"

	"gogobee/internal/db"
	"gogobee/internal/util"

	"maunium.net/go/mautrix"
	"maunium.net/go/mautrix/id"
)

// levelUpMessages are Twinbee/Parodius themed congratulations.
var levelUpMessages = []string{
	"Power-up collected! %s just reached Level %d! 🔔",
	"Twinbee's bell is ringing! %s ascended to Level %d! 🛎️",
	"Parodius would be proud — %s hit Level %d! 🐙",
	"Stage clear! %s leveled up to Level %d! ⭐",
	"Winbee sends sparkles — %s is now Level %d! ✨",
	"Bonus round complete! %s reached Level %d! 🎰",
	"GwinBee detected a power surge — %s is Level %d! ⚡",
	"The Vic Viper salutes %s for reaching Level %d! 🚀",
	"Takosuke is dancing! %s just became Level %d! 🎶",
	"A shower of bells for %s — Level %d unlocked! 🔔🔔🔔",
	"Pastel confirms: %s has broken through to Level %d! 🌈",
	"The Shooting Star squadron welcomes %s to Level %d! 💫",
}

// XPPlugin awards XP for messages and tracks levels.
type XPPlugin struct {
	Base
	mu        sync.Mutex
	cooldowns map[id.UserID]time.Time
}

// NewXPPlugin creates a new XP plugin.
func NewXPPlugin(client *mautrix.Client) *XPPlugin {
	return &XPPlugin{
		Base:      NewBase(client),
		cooldowns: make(map[id.UserID]time.Time),
	}
}

func (p *XPPlugin) Name() string { return "xp" }

func (p *XPPlugin) Commands() []CommandDef {
	return []CommandDef{
		{Name: "rank", Description: "Show your level, XP, and progress", Usage: "!rank [@user]", Category: "Leveling & Stats"},
		{Name: "leaderboard", Description: "Show top 10 users by XP", Usage: "!leaderboard", Category: "Leveling & Stats"},
	}
}

func (p *XPPlugin) Init() error { return nil }

func (p *XPPlugin) OnReaction(_ ReactionContext) error { return nil }

func (p *XPPlugin) OnMessage(ctx MessageContext) error {
	// Handle commands
	if p.IsCommand(ctx.Body, "rank") {
		return p.handleRank(ctx)
	}
	if p.IsCommand(ctx.Body, "leaderboard") {
		return p.handleLeaderboard(ctx)
	}

	// Skip XP for commands
	if ctx.IsCommand {
		return nil
	}

	// Passive XP with cooldown
	p.mu.Lock()
	last, ok := p.cooldowns[ctx.Sender]
	now := time.Now().UTC()
	if ok && now.Sub(last) < 30*time.Second {
		p.mu.Unlock()
		return nil
	}
	p.cooldowns[ctx.Sender] = now
	// Periodically clean up expired cooldowns to prevent unbounded growth
	if len(p.cooldowns) > 1000 {
		for uid, t := range p.cooldowns {
			if now.Sub(t) > time.Minute {
				delete(p.cooldowns, uid)
			}
		}
	}
	p.mu.Unlock()

	_, leveledUp, newLevel := p.grantXPInternal(ctx.Sender, 10, "message")
	if leveledUp {
		msg := levelUpMessages[rand.Intn(len(levelUpMessages))]
		announcement := fmt.Sprintf(msg, string(ctx.Sender), newLevel)
		if err := p.SendMessage(ctx.RoomID, announcement); err != nil {
			slog.Error("failed to send level-up announcement", "err", err)
		}
	}

	return nil
}

// GrantXP awards XP to a user from an external plugin and returns (newXP, leveledUp, newLevel).
func (p *XPPlugin) GrantXP(userID id.UserID, amount int, reason string) (int, bool, int) {
	return p.grantXPInternal(userID, amount, reason)
}

func (p *XPPlugin) grantXPInternal(userID id.UserID, amount int, reason string) (int, bool, int) {
	d := db.Get()
	now := time.Now().UTC().Unix()

	// Ensure user row exists
	_, err := d.Exec(
		`INSERT INTO users (user_id, xp, level, last_xp_at) VALUES (?, 0, 0, 0)
		 ON CONFLICT(user_id) DO NOTHING`, string(userID))
	if err != nil {
		slog.Error("xp: ensure user", "err", err)
		return 0, false, 0
	}

	// Get current state
	var oldXP, oldLevel int
	err = d.QueryRow(`SELECT xp, level FROM users WHERE user_id = ?`, string(userID)).Scan(&oldXP, &oldLevel)
	if err != nil {
		slog.Error("xp: read user", "err", err)
		return 0, false, 0
	}

	newXP := oldXP + amount
	newLevel := util.LevelFromXP(newXP)
	leveledUp := newLevel > oldLevel

	_, err = d.Exec(
		`UPDATE users SET xp = ?, level = ?, last_xp_at = ? WHERE user_id = ?`,
		newXP, newLevel, now, string(userID))
	if err != nil {
		slog.Error("xp: update user", "err", err)
		return oldXP, false, oldLevel
	}

	// Log the XP grant
	_, err = d.Exec(
		`INSERT INTO xp_log (user_id, amount, reason) VALUES (?, ?, ?)`,
		string(userID), amount, reason)
	if err != nil {
		slog.Error("xp: log grant", "err", err)
	}

	return newXP, leveledUp, newLevel
}

func (p *XPPlugin) handleRank(ctx MessageContext) error {
	target := ctx.Sender
	args := p.GetArgs(ctx.Body, "rank")
	if args != "" {
		if resolved, ok := p.ResolveUser(args, ctx.RoomID); ok {
			target = resolved
		}
	}

	d := db.Get()
	var xp, level int
	err := d.QueryRow(`SELECT xp, level FROM users WHERE user_id = ?`, string(target)).Scan(&xp, &level)
	if err == sql.ErrNoRows {
		return p.SendReply(ctx.RoomID, ctx.EventID, fmt.Sprintf("%s hasn't earned any XP yet.", string(target)))
	}
	if err != nil {
		slog.Error("xp: rank query", "err", err)
		return p.SendReply(ctx.RoomID, ctx.EventID, "Failed to look up rank.")
	}

	nextLevel := level + 1
	xpForNext := util.XPForLevel(nextLevel)
	xpForCurrent := util.XPForLevel(level)
	progress := xp - xpForCurrent
	needed := xpForNext - xpForCurrent
	bar := util.ProgressBar(progress, needed, 20)

	// Get rank position (users with same XP share the same rank)
	var rank int
	err = d.QueryRow(`SELECT COUNT(DISTINCT xp) + 1 FROM users WHERE xp > ?`, xp).Scan(&rank)
	if err != nil {
		rank = 0
	}

	msg := fmt.Sprintf(
		"📊 %s — Level %d (Rank #%s)\nXP: %s / %s\n%s",
		string(target), level, formatNumber(rank),
		formatNumber(xp), formatNumber(xpForNext), bar,
	)
	return p.SendReply(ctx.RoomID, ctx.EventID, msg)
}

func (p *XPPlugin) handleLeaderboard(ctx MessageContext) error {
	members := p.RoomMembers(ctx.RoomID)

	d := db.Get()
	rows, err := d.Query(`SELECT user_id, xp, level FROM users ORDER BY xp DESC`)
	if err != nil {
		slog.Error("xp: leaderboard query", "err", err)
		return p.SendReply(ctx.RoomID, ctx.EventID, "Failed to load leaderboard.")
	}
	defer rows.Close()

	var sb strings.Builder
	sb.WriteString("🏆 XP Leaderboard — Top 10\n\n")

	medals := []string{"🥇", "🥈", "🥉"}
	i := 0
	for rows.Next() && i < 10 {
		var userID string
		var xp, level int
		if err := rows.Scan(&userID, &xp, &level); err != nil {
			continue
		}
		if members != nil && !members[id.UserID(userID)] {
			continue
		}
		prefix := fmt.Sprintf("#%d", i+1)
		if i < len(medals) {
			prefix = medals[i]
		}
		sb.WriteString(fmt.Sprintf("%s %s — Level %d (%s XP)\n", prefix, userID, level, formatNumber(xp)))
		i++
	}

	if i == 0 {
		return p.SendReply(ctx.RoomID, ctx.EventID, "No XP data yet.")
	}

	return p.SendReply(ctx.RoomID, ctx.EventID, sb.String())
}

// formatNumber adds commas to an integer for display.
func formatNumber(n int) string {
	if n < 0 {
		return "-" + formatNumber(-n)
	}
	s := fmt.Sprintf("%d", n)
	if len(s) <= 3 {
		return s
	}
	var result strings.Builder
	remainder := len(s) % 3
	if remainder > 0 {
		result.WriteString(s[:remainder])
	}
	for i := remainder; i < len(s); i += 3 {
		if result.Len() > 0 {
			result.WriteByte(',')
		}
		result.WriteString(s[i : i+3])
	}
	return result.String()
}
